src/Compilers/Test/Core/Compilation/OperationTreeVerifier.cs

```diff
@@ -62,13 +63,186 @@ public static string GetOperationTree(Compilation compilation, IOperation operat
+                        i = buffer.AsSpan().Slice(0, i).LastIndexOf('\n');
+                        if (i < 0) break;
+
+                        // Check for lack of preceding \r
+                        if (i == 0 || buffer[i - 1] != '\r')
+                        {
+                            // Expand / borrow toInsert array if required
+                            toInsert ??= ArrayPool<int>.Shared.Rent(64);
+                            if (toInsertCount == toInsert.Length)
+                            {
+                                var newArr = ArrayPool<int>.Shared.Rent(toInsert.Length * 2);
+                                toInsert.AsSpan().CopyTo(newArr.AsSpan().Slice(0, toInsert.Length));
+                                ArrayPool<int>.Shared.Return(toInsert);
+                                toInsert = newArr;
+                            }
+
+                            // Write value into removal buffer
+                            toInsert[toInsertCount++] = i;
+                        }
+                    }
+
+                    // Insert \r at all these positions
+                    InsertHelper(ref buffer, ref bufferSize, toInsert, toInsertCount, '\r');
+                }
+                else if (Environment.NewLine == "\n")
+                {
+                    // Find all CRs, and add them to intBuffer (our toRemove list)
+                    ref int[] toRemove = ref intBuffer;
+                    int toRemoveCount = 0;
+                    int i = bufferSize;
+                    while (true)
+                    {
+                        // Go to next preceding \r (if any)
+                        i = buffer.AsSpan().Slice(0, i).LastIndexOf('\r');
+                        if (i < 0) break;
+
+                        // Expand / borrow toRemove array if required
+                        toRemove ??= ArrayPool<int>.Shared.Rent(64);
+                        if (toRemoveCount == toRemove.Length)
+                        {
+                            var newArr = ArrayPool<int>.Shared.Rent(toRemove.Length * 2);
+                            toRemove.AsSpan().CopyTo(newArr.AsSpan().Slice(0, toRemove.Length));
+                            ArrayPool<int>.Shared.Return(toRemove);
+                            toRemove = newArr;
+                        }
+
+                        // Write value into removal buffer
+                        toRemove[toRemoveCount++] = i;
+                    }
+
+                    // Perform removal
+                    RemoveHelper(ref buffer, ref bufferSize, toRemove, toRemoveCount);
+                }
+                else
+                {
+                    throw new NotSupportedException("Only supported line endings are lf and crlf (OperationTreeVerifier.Verify)");
+                }
+            }
+
+            // Emulate actualOperationTree.Trim(newLineChars).Replace(" \n", "\n").Replace(" \r", "\r") into actualBuffer with actualBufferSize
```

Comment:
I don't think I have the brain power to review this today, but I'm going to insist that we actually spell out the general algorithm going on here in more plain english.